### PR TITLE
cleanup(pubsub): remove unused code; tidy up other code; edit documentation

### LINKED ISSUE #6306: Convert google/cloud/pubsub to use g::c::Options
Change the library to *prefer* `google::cloud::Options` in its initialization, the overloads taking `google::cloud::pubsub::ClientOptions` should remain usable for backwards compatibility.

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #6306 

The only remaining tasks are captured by: #7344, #7347 , #7348

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7350)

### CODE REVIEW

File: google/cloud/pubsub/internal/defaults.cc [hunk 1]
```diff
@@ -61,10 +60,14 @@ Options DefaultCommonOptions(Options opts) {
     opts.set<GrpcTracingOptionsOption>(internal::DefaultTracingOptions());
   }
   if (!opts.has<pubsub::RetryPolicyOption>()) {
-    opts.set<pubsub::RetryPolicyOption>(DefaultRetryPolicy()->clone());
+    opts.set<pubsub::RetryPolicyOption>(
+        std::make_shared<pubsub::LimitedTimeRetryPolicy>(
```
Comment (devbww) [inline google/cloud/pubsub/internal/defaults.cc:64]: It might be worth sticking with the `clone()` idiom rather than using `make_shared` directly.

**Review by coryan — APPROVED** [review]

File: google/cloud/pubsub/internal/defaults.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/pubsub/internal/defaults.cc]: Done. Maybe you can explain to me why we might prefer one over the other?

Comment (devbww) [reply google/cloud/pubsub/internal/defaults.cc]: The big difference is that `clone()` is polymorphic.  While that doesn't matter in these particular cases, I reckoned that if we have `clone()` anyway then we might as well use it, even when we know the compile-time type.

Aside: I think you missed a couple of occurrences in the `test_retry_policies.cc` edits, but not worth worrying about now I'd proffer.

Comment (dbolduc) [reply google/cloud/pubsub/internal/defaults.cc]: yup, I missed those 🤦‍♂️

**Pull request merged**